src/ui/src/channel/message_list.rs

```diff
@@ -1,50 +1,65 @@
 use gpui::{div, IntoElement, ListAlignment, ListState, ParentElement, Pixels};
 
-use scope_chat::message::Message;
+use scope_chat::message::{Message, MessageAuthor};
 
-use super::message::message;
+use super::message::{message, MessageGroup};
 
 #[derive(Clone)]
 pub struct MessageList<M: Message + 'static> {
-  real_messages: Vec<M>,
-  pending_messages: Vec<M>,
+  messages: Vec<MessageGroup<M>>,
 }
 
 impl<M: Message> MessageList<M> {
   pub fn new() -> MessageList<M> {
-    Self {
-      real_messages: Vec::default(),
-      pending_messages: Vec::default(),
-    }
+    Self { messages: Vec::default() }
   }
 
   pub fn add_external_message(&mut self, message: M) {
-    if let Some((_, pending_index)) = self
-      .pending_messages
-      .iter()
-      .zip(0..)
-      .find(|(msg, _)| msg.get_nonce().map(|v1| message.get_nonce().map(|v2| v2 == v1).unwrap_or(false)).unwrap_or(false))
-    {
-      self.pending_messages.remove(pending_index);
+    if let Some(nonce) = message.get_nonce() {
+      let mut removal_index: Option<usize> = None;
+
+      for (group, index) in self.messages.iter_mut().zip(0..) {
+        let matching = group.find_matching(nonce);
+
+        if let Some(matching) = matching {
+          if group.size() == 1 {
+            removal_index = Some(index);
+          } else {
+            group.remove(matching);
+          }
+        }
+      }
+
+      if let Some(removal_index) = removal_index {
+        self.messages.remove(removal_index);
+      }
     }
 
-    self.real_messages.push(message);
+    let last = self.messages.last_mut();
+
+    if last.is_some() && last.as_ref().unwrap().get_author().get_id() == message.get_author().get_id() {
+      last.unwrap().add(message);
+    } else {
+      self.messages.push(MessageGroup::new(message));
+    }
```

Comment:
Could we add a check for if the last message was over 5 minutes ago? Discord will ungroup in that case

Reply:
aaah, didn't think about that. Will do now